crates/astria-sequencer-relayer/src/relayer.rs

```diff
@@ -161,21 +344,129 @@ impl Relayer {
+        }
+        // .await
+        // .wrap_err("failed establishing connection to data availability layer")?;
+        self.wait_for_sequencer(5, Duration::from_secs(5), 2.0)
+            .await
+            .wrap_err("failed establishing connection to the sequencer")?;
+
+        let mut sequencer_interval = interval(self.sequencer_poll_period);
+        // the first tick resolves immediately, so do it now because we don't
+        // want to poll right after start.
+        sequencer_interval.tick().await;
+
+        let stop_msg = loop {
+            select!(
+                // Query sequencer for the latest block if no task is in flight
+                _ = sequencer_interval.tick() => self.handle_sequencer_tick(),
+
+                // Handle the sequencer response by converting it
+                //
+                // NOTE: + wrapping the task in an async block makes this lazy;
+                //       + `unwrap`ping can't fail because this branch is disabled if `None`
+                res = async { self.sequencer_task.as_mut().unwrap().await }, if self.sequencer_task.is_some() => {
+                    self.handle_sequencer_response(res);
+                }
+
+                // Distribute and store converted/admitted blocks
+                Some(res) = self.conversion_workers.join_next() => {
+                    if self.handle_conversion_completed(res)
+                           .is_gossip_channel_closed()
+                    {
+                        break "gossip block channel closed unexpectedly";
+                    }
+                }
+
+                // Record the current height of the data availability layer if a submission
+                // was in flight.
+                //
+                // NOTE: + wrapping the task in an async block makes this lazy;
+                //       + `unwrap`ping can't fail because this branch is disabled if `None`
+                res = async { self.submission_task.as_mut().unwrap().await }, if self.submission_task.is_some() => {
+                    self.handle_submission_completed(res);
                 }
-                Ok(_) => {}
+            );
+            // Try to submit new blocks
+            //
+            // This will immediately and eagerly try to submit to the data availability
+            // layer if no submission is in flight.
+            if self.data_availability.is_some()
+                && !self.queued_blocks.is_empty()
+                && self.submission_task.is_none()
+            {
+                let client = self.data_availability.clone().expect(
+                    "this should not fail because the if condition of this block checked that a \
+                     client is present",
+                );
+                self.submission_task = Some(task::spawn(submit_blocks_to_data_availability_layer(
+                    client,
+                    self.queued_blocks.clone(),
```

Comment:
this seems pricey - can we just take all the values out of the vec or something?